File: spier-datalog/src/stats.rs
/// Read-only schema and cardinality abstraction used by the compiler/planner
/// for cost estimation. Implementors are typically adapters over the transactor.
pub trait CompileStats: Send + Sync {
    fn lookup_attr(&self, name: &str) -> Option<u32>;

    fn estimate_index_size(&self, index: &str, bound: &[u64]) -> f64;

    fn partition_id_for(&self, name: &str) -> Option<u64>;

    fn is_ref_attr(&self, attr_name: &str) -> bool;

    fn is_indexed_attr(&self, attr_name: &str) -> bool;
}

File: nim-datalog/stats.nim
type
  CompileStats* = object
    lookupAttr*: proc(name: string): uint32 {.closure.}
    estimateIndexSize*: proc(index: string, bound: openArray[uint64]): float64 {.closure.}
    partitionIdFor*: proc(name: string): uint64 {.closure.}
    isRefAttr*: proc(name: string): bool {.closure.}
    isIndexedAttr*: proc(name: string): bool {.closure.}
